8289607: Change hotspot/jtreg tests to not use Thread.suspend/resume

lmesnik opened the pull request:
The Thread.suspend/resume is replaced by JVM TI  SuspendThread/ResumeThread whether it is possible. 
Testcase from test  test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/interrupt/interrupt001.java was removed because only one environment can suspend threads so debugger is not compatible with JVM TI agent suspending test.

Test test/hotspot/jtreg/runtime/handshake/HandshakeSuspendExitTest.java starts failing because it got JVMTI_ERROR_INTERNAL. The problem is that JvmtiSuspendControl::resume() returned false. However, it might happen only if the thread is not suspended, which means that it should be JVMTI_ERROR_THREAD_NOT_SUSPENDED.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8289607](https://bugs.openjdk.org/browse/JDK-8289607): Change hotspot/jtreg tests to not use Thread.suspend/resume


### Reviewers
 * [Serguei Spitsyn](https://openjdk.org/census#sspitsyn) (@sspitsyn - **Reviewer**)
 * [Chris Plummer](https://openjdk.org/census#cjplummer) (@plummercj - **Reviewer**) ⚠️ Review applies to [c892fca0](https://git.openjdk.org/jdk/pull/10351/files/c892fca0cbbde47758195ff9cac050228f58f646)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10351/head:pull/10351` \
`$ git checkout pull/10351`

Update a local copy of the PR: \
`$ git checkout pull/10351` \
`$ git pull https://git.openjdk.org/jdk pull/10351/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10351`

View PR using the GUI difftool: \
`$ git pr show -t 10351`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10351.diff">https://git.openjdk.org/jdk/pull/10351.diff</a>

</details>

sspitsyn commented on src/hotspot/share/prims/jvmtiEnvBase.cpp:
```diff
@@ -1666,7 +1666,7 @@ JvmtiEnvBase::resume_thread(oop thread_oop, JavaThread* java_thread, bool single
     if (java_thread->is_suspended()) {
       if (!JvmtiSuspendControl::resume(java_thread)) {
-        return JVMTI_ERROR_INTERNAL;
+        return JVMTI_ERROR_THREAD_NOT_SUSPENDED;
```
This is right.

sspitsyn commented on test/hotspot/jtreg/testlibrary/jvmti/libJvmtiUtils.cpp:
```diff
@@ -52,10 +53,34 @@ JNIEXPORT void JNICALL
+    return;
+  }
+  check_jvmti_status(jni, err, "Error during ResumeThread");
+}
```
Would it better to return the jvmtiError status as an int, so it can be checked at the java level?
In fact, it is not good that the `JVMTI_ERROR_THREAD_NOT_ALIVE`, `JVMTI_ERROR_THREAD_SUSPENDED` and `JVMTI_ERROR_THREAD_NOT_SUSPENDED` are always ignored.

plummercj replied:
I was going to ask something similar. Should the tests ever see these errors? If not, probably an exception should be thrown here. If they are ok in some situations, then the tests should be checking for them.

sspitsyn commented on test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/interrupt/interrupt001.java:
```diff
@@ -133,7 +133,7 @@ private static void log3(String message) {
     static final String EQUALS_NOT_INTERRUPTED = "Statuses of threads are equal: not interrupted";
 
-    static final int MAX_CASE = 2;
+    static final int MAX_CASE = 1;
```
This constant is not used anymore. Should it be deleted?

sspitsyn reviewed (commented):
Hi Leonid,
This looks good in general.
I've added a couple of comments though.
The PR title does not match the bug title: com/sun/jdi vs hotspot/jtreg.
Thanks,
Serguei

plummercj commented:
>  Testcase from test test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/interrupt/interrupt001.java was removed because only one environment can suspend threads so debugger is not compatible with JVM TI agent suspending test.

As far as I know, this is the only test we have for ThreadReference.interrupt(). Also see [JDK-8282384](https://bugs.openjdk.org/browse/JDK-8282384), which points out that we have no test for ThreadReference.interrupt() of a virtual thread, and suggests modifying this test to support virtual threads. I actually attempted to do this, but ran into an issue with ThreadReference.resume() support for virtual threads, although I'm not sure what that issue was, and it may be fixed now.

lmesnik commented:
> > Testcase from test test/hotspot/jtreg/vmTestbase/nsk/jdi/ThreadReference/interrupt/interrupt001.java was removed because only one environment can suspend threads so debugger is not compatible with JVM TI agent suspending test.
> 
> As far as I know, this is the only test we have for ThreadReference.interrupt(). Also see [JDK-8282384](https://bugs.openjdk.org/browse/JDK-8282384), which points out that we have no test for ThreadReference.interrupt() of a virtual thread, and suggests modifying this test to support virtual threads. I actually attempted to do this, but ran into an issue because Thread.resume() is not supported for virtual threads, but seems your changes get rid of resume() so it shouldn't be an issue anymore.
> 
> I think losing this test case means we no longer test interrupt when suspended. Maybe we need to have the debugger side initiate the suspending.

Let me file another bug for this. Just don't block Alan's fix https://github.com/openjdk/jdk/pull/10324

plummercj reviewed (approved):
serviceability changes look good

sspitsyn commented:
Leonid,
I've posted a couple of comments but am not sure you see them.
Do you see my comments? If so, are you going to address them?

lmesnik commented:
> Leonid, I've posted a couple of comments but am not sure you see them. Do you see my comments? If so, are you going to address them?

Yes, I updated JvmtiUtils and removed unused variable. The PR title is correct now.

sspitsyn reviewed (approved):
Looks good.
Thank you for the update.
Serguei

lmesnik commented:
/integrate
